use less stack on Print::printNumber

DEvil0000 opened the pull request:
The issue with printNumber is that it uses a lot of stack quite unexpectedly. This makes it basically unusable on a attiny.
I suggest doing the naive implementation as it's done with float print. 

I think the Autor was hoping for a more optimized write call Implementation in the over-writable write call with a long buffer. I think this is too late to get it more efficient for most platforms anyway. Another option of course would be changing the API and adding a parameter for this or adding a define to switch the default Implementation.

DEvil0000 committed:
use less stack on number print

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/arduino/ArduinoCore-avr?pullRequest=528) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/arduino/ArduinoCore-avr?pullRequest=528) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/arduino/ArduinoCore-avr?pullRequest=528) it.</sub>

per1234 commented:
Hi @DEvil0000. Thanks for your pull request. I'm just commenting to let you know the failure of the "Spell Check" workflow run is not related to the changes you proposed (it was caused by a recent automated update of the spellcheck dictionary) and won't affect the review of the PR.

I have now fixed the problem in the `master` branch: https://github.com/arduino/ArduinoCore-avr/pull/529

PaulStoffregen commented:
This PR causes digits to print in reverse order.

Tested with Arduino Uno using IDE 1.8.19 with AVR core 1.8.6.

![screenshot](https://user-images.githubusercontent.com/965463/229434063-8349d60b-faa4-47b2-acdb-db85cae1a6eb.png)

![screenshot2](https://user-images.githubusercontent.com/965463/229434228-1245dde5-94a4-4afc-8fd7-4e2a65c7938e.png)

DEvil0000 committed:
fixed for 10 digit numbers now working for full long int range

DEvil0000 committed:
improved on base 2 and base 16 print number implementations
added compile flag to disable those additional optimized implementations

DEvil0000 commented:
@PaulStoffregen sorry, looks like I was a bit sleepy on the first PR. I fixed the implementation with the first two commits and added improved versions for base 2 and base 16 which one can disable by compiler define flag.

DEvil0000 committed:
fixed comment

DEvil0000 committed:
fixed comment

DEvil0000 committed:
fixed printNumber base2 and base16 code for -mint8

DEvil0000 committed:
avoid writing to base variable so the compiler has a chance of better optimization due to it being basically const then

TBAMax commented:
Yeah, catering for all the bases with one function messes it up there. When printing decimal only, much smaller buffer is needed(11bytes vs 33bytes).
The good formula for calculating needed digits for decimal:
`#define MAX_DECIMAL_SIZE(x)  ((size_t)(8 * sizeof(x) * 302 / 1000) + 1)`
